pageserver/src/page_service.rs

```diff
@@ -990,27 +991,64 @@ impl postgres_backend_async::Handler for PageServerHandler {
     }
 }
 
+#[derive(thiserror::Error, Debug)]
+enum GetActiveTenantError {
+    #[error(
+        "Timed out waiting {wait_time:?} for tenant active state. Latest state: {latest_state:?}"
+    )]
+    WaitForActiveTimeout {
+        latest_state: TenantState,
+        wait_time: Duration,
+    },
+    #[error(transparent)]
+    Other(#[from] anyhow::Error),
+}
+
+impl From<GetActiveTenantError> for QueryError {
+    fn from(e: GetActiveTenantError) -> Self {
+        match e {
+            GetActiveTenantError::WaitForActiveTimeout { .. } => QueryError::Disconnected(
+                ConnectionError::Socket(io::Error::new(io::ErrorKind::TimedOut, e.to_string())),
+            ),
+            GetActiveTenantError::Other(e) => QueryError::Other(e),
+        }
+    }
+}
+
 /// Get active tenant.
 ///
 /// If the tenant is Loading, waits for it to become Active, for up to 30 s. That
 /// ensures that queries don't fail immediately after pageserver startup, because
 /// all tenants are still loading.
-async fn get_active_tenant_with_timeout(tenant_id: TenantId) -> anyhow::Result<Arc<Tenant>> {
+async fn get_active_tenant_with_timeout(
+    tenant_id: TenantId,
```

Comment:
shouldn't the mgr expose this functionality?

Reply:
`mgr` has similar notion of returning active tenants only, when asked: https://github.com/neondatabase/neon/blob/5223b62a194269bc5d19d325eacc357f48ed9302/pageserver/src/tenant/mgr.rs#L248

Generally, we need the timeouts to wait during client request serving, during attaching a certain tenant with many timelines.
For now, something similar to this is done in `tasks.rs` only, where GC and compaction wait for active tenant to start running.
I don't have an answer why this is done so, really, it looks to me that only one approach could be left, but maybe somebody else could know.